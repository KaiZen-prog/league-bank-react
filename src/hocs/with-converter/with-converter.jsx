import { connect } from 'react-redux';
import { addConversion, changeDate } from '../../store/actions';
import moment from 'moment';
import React from 'react';
import '../../../node_modules/react-datepicker/dist/react-datepicker.css';
import { loadExchangeRate } from '../../store/api-actions';
import { FormFields, Currencies, FLOAT_COEFFICIENT } from '../../const';
import PropTypes from 'prop-types';

export const withConverter = (Component) => {
  class WithCurrencyConverter extends React.PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        exchangeRate: props.exchangeRate,
        currencyInput: {
          type: Currencies.RUB,
          amount: 0,
        },
        currencyOutput: {
          type: Currencies.USD,
          amount: 0,
        },
      };

      this.submitHandler = this.submitHandler.bind(this);
      this.typeChangeHandler = this.typeChangeHandler.bind(this);
      this.valueConversion = this.valueConversion.bind(this);
      this.conversionFromUSD = this.conversionFromUSD.bind(this);
      this.conversionToUSD = this.conversionToUSD.bind(this);
      this.valueChangeHandler = this.valueChangeHandler.bind(this);
      this.dateChangeHandler = this.dateChangeHandler.bind(this);
    }

    componentDidMount() {
      this.exchangeRateUpdate(this.props.date);
    }

    submitHandler(evt) {
      evt.preventDefault();
      this.props.addTransaction({
        date: moment(this.props.date).format('DD.MM.YYYY'),
        currencyInput: this.state.currencyInput,
        currencyOutput: this.state.currencyOutput,
      });
    }

    typeChangeHandler(evt) {
      const { name, value } = evt.target;
      const inputAmount = this.state.currencyInput.amount;

      if (value === this.state.currencyInput.type || value === this.state.currencyOutput.type) {
        switch (name) {
          case FormFields.INPUT: {
            this.setState({
              currencyInput: {
                type: value,
                amount: inputAmount,
              },
              currencyOutput: {
                type: value,
                amount: inputAmount,
              },
            });
            return;
          }

          case FormFields.OUTPUT: {
            this.setState({
              currencyOutput: {
                type: value,
                amount: inputAmount,
              },
            });
            return;
          }
        }
      }

      this.setState(
        (prevState) => ({
          [name]: Object.assign({}, prevState[name], { type: value }),
        }),
        () => {
          this.valueConversion(FormFields.INPUT, inputAmount);
        },
      );
    }

    conversionToUSD(name, value) {
      return (
        Math.floor((value / this.state.exchangeRate[this.state[name].type]) * FLOAT_COEFFICIENT) /
        FLOAT_COEFFICIENT
      );
    }

    conversionFromUSD(name, value) {
      return (
        Math.floor(value * this.state.exchangeRate[this.state[name].type] * FLOAT_COEFFICIENT) /
        FLOAT_COEFFICIENT
      );
    }

    valueConversion(name, value) {
      if (name === FormFields.INPUT) {
        this.entryField = FormFields.INPUT;
        this.outputField = FormFields.OUTPUT;
      } else {
        this.entryField = FormFields.OUTPUT;
        this.outputField = FormFields.INPUT;
      }

      const convertedToUSD = this.conversionToUSD(this.entryField, value);
      const result = this.conversionFromUSD(this.outputField, convertedToUSD);

      this.setState((prevState) => ({
        [this.outputField]: Object.assign({}, prevState[this.outputField], { amount: result }),
      }));
    }

    valueChangeHandler(evt) {
      const { name, value } = evt.target;
      this.setState((prevState) => ({
        [name]: Object.assign({}, prevState[name], {
          amount: value === '' ? '' : Math.floor(value * FLOAT_COEFFICIENT) / FLOAT_COEFFICIENT,
        }),
      }));

      this.valueConversion(name, value);
    }

    exchangeRateUpdate(date) {
      this.props.loadExchangeRate(date, () => {
        this.setState({ exchangeRate: this.props.exchangeRate }, () => {
          this.valueConversion(FormFields.INPUT, this.state[FormFields.INPUT].amount);
        });
      });
    }

    dateChangeHandler(date) {
      const formatDate = moment(date).format('YYYY-MM-DD');
      this.props.changeDate(formatDate);
      this.exchangeRateUpdate(formatDate);
    }

    render() {
      const state = this.state;

      return (
        <Component
          state={state}
          date={this.props.date}
          onChange={(date) => this.dateChangeHandler(date)}
          submitHandler={this.submitHandler}
          typeChangeHandler={this.typeChangeHandler}
          valueChangeHandler={this.valueChangeHandler}
        >
        </Component>
      );
    }
  }

  WithCurrencyConverter.propTypes = {
    date: PropTypes.string,
    exchangeRate: PropTypes.shape({
      USD: PropTypes.number.isRequired,
      RUB: PropTypes.number.isRequired,
      EUR: PropTypes.number.isRequired,
      GBP: PropTypes.number.isRequired,
      CNY: PropTypes.number.isRequired,
    }).isRequired,
    addTransaction: PropTypes.func.isRequired,
    changeDate: PropTypes.func.isRequired,
    loadExchangeRate: PropTypes.func.isRequired,
  };

  const mapStateToProps = (state) => ({
    date: state.CONVERTER.date,
    exchangeRate: state.CONVERTER.exchangeRate,
  });

  const mapDispatchToProps = (dispatch) => ({
    addTransaction(transaction) {
      dispatch(addConversion(transaction));
    },

    changeDate(date) {
      dispatch(changeDate(date));
    },

    loadExchangeRate(date, callback) {
      dispatch(loadExchangeRate(date, callback));
    },
  });

  return connect(mapStateToProps, mapDispatchToProps)(WithCurrencyConverter);
};
