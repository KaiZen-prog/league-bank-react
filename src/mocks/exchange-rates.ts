export const exchangeRates = {
  '2024-01-19': {
    USD: 1,
    RUB: 88.573959,
    EUR: 0.918438,
    KZT: 452.633544,
    GBP: 0.787512,
    CNY: 7.1197
  },
  '2024-01-18': {
    USD: 1,
    RUB: 88.638497,
    EUR: 0.918649,
    KZT: 453.345273,
    GBP: 0.788835,
    CNY: 7.1263
  },
  '2024-01-17': {
    USD: 1,
    RUB: 88.105727,
    EUR: 0.919239,
    KZT: 453.139857,
    GBP: 0.791321,
    CNY: 7.1265
  },
  '2024-01-16': {
    USD: 1,
    RUB: 87.719298,
    EUR: 0.913662,
    KZT: 451.339299,
    GBP: 0.786436,
    CNY: 7.1073
  },
  '2024-01-15': {
    USD: 1,
    RUB: 88.043653,
    EUR: 0.913639,
    KZT: 451.160232,
    GBP: 0.785191,
    CNY: 7.111
  },
  '2024-01-14': {
    USD: 1,
    RUB: 87.959997,
    EUR: 0.912076,
    KZT: 451.160232,
    GBP: 0.78456,
    CNY: 7.1126
  },
  '2024-01-13': {
    USD: 1,
    RUB: 87.959997,
    EUR: 0.912076,
    KZT: 451.530905,
    GBP: 0.78456,
    CNY: 7.1106
  },
  '2024-01-12': {
    USD: 1,
    RUB: 88.495575,
    EUR: 0.910544,
    KZT: 451.995747,
    GBP: 0.782515,
    CNY: 7.09752
  }
};
