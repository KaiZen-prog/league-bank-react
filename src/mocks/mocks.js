export const promoSlides = [
  `img/desktop-slide-1.jpg`,
  `img/desktop-slide-2.jpg`,
  `img/desktop-slide-3.jpg`,
];
