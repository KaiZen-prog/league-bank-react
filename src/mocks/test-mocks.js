import moment from "moment";

export const InitialState = {
  CONVERSION_HISTORY: {
    conversionHistory: [
      {
        date: `18.09.2021`,
        currencyInput: {
          type: `USD`,
          amount: 6549
        },
        currencyOutput: {
          type: `EUR`,
          amount: 5585.07
        }
      },
      {
        date: `19.09.2021`,
        currencyInput: {
          type: `CNY`,
          amount: 23
        },
        currencyOutput: {
          type: `RUB`,
          amount: 258.59
        }
      },
      {
        date: `20.09.2021`,
        currencyInput: {
          type: `EUR`,
          amount: 12458
        },
        currencyOutput: {
          type: `GBP`,
          amount: 10694.16
        }
      },
      {
        date: `13.09.2021`,
        currencyInput: {
          type: `CNY`,
          amount: 654
        },
        currencyOutput: {
          type: `RUB`,
          amount: 7363.93
        }
      },
      {
        date: `15.09.2021`,
        currencyInput: {
          type: `EUR`,
          amount: 105
        },
        currencyOutput: {
          type: `GBP`,
          amount: 89.6
        }
      },
      {
        date: `20.09.2021`,
        currencyInput: {
          type: `RUB`,
          amount: 105
        },
        currencyOutput: {
          type: `USD`,
          amount: 1.43
        }
      }
    ]
  },

  CONVERTER: {
    date: moment().utc().format(`YYYY-MM-DD`),
    exchangeRate: {
      USD: 1,
      RUB: 72.8797,
      EUR: 0.852813,
      GBP: 0.727776,
      CNY: 6.4662
    }
  }
};
