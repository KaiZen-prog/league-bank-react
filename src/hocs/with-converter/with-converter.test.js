import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from "history";
import {withConverter} from "./with-converter";
import {InitialState} from '../../mocks/test-mocks';

const mockStore = configureStore();
const history = createMemoryHistory({initialEntries: [`/`]});

const noop = () => {};

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

const MockComponentWrapped = withConverter(MockComponent);

it(`withConverter is rendered correctly`, () => {
  const store = mockStore(InitialState);

  const tree = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <MockComponentWrapped
            date={InitialState.CONVERTER.date}
            exchangeRate={InitialState.CONVERTER.exchangeRate}
            addConversion={noop}
            changeDate={noop}
            loadExchangeRate={noop}
          />
        </Router>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

MockComponent.propTypes = {
  children: PropTypes.node,
};
