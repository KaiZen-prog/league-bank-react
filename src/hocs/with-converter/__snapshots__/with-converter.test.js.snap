// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withConverter is rendered correctly 1`] = `<div />`;
